src/internet_identity/src/archive.rs

```diff
@@ -0,0 +1,247 @@
+use crate::state;
+use crate::state::{ArchiveData, ArchiveState, ArchiveStatusCache, DeviceDataInternal};
+use candid::Principal;
+use ic_cdk::api::call::{call_with_payment, CallResult};
+use ic_cdk::api::management_canister::main::{
+    canister_status, install_code, CanisterIdRecord, CanisterInstallMode,
+    CanisterInstallMode::Install, CanisterStatusResponse, CreateCanisterArgument,
+    InstallCodeArgument,
+};
+use ic_cdk::api::time;
+use ic_cdk::{id, notify};
+use internet_identity_interface::{
+    ArchiveInit, DeployArchiveResult, DeviceDataUpdate, Entry, Operation, Private, UserNumber,
+};
+use serde_bytes::ByteBuf;
+use sha2::Digest;
+use sha2::Sha256;
+use std::time::Duration;
+use ArchiveState::{Created, CreationInProgress, NotCreated};
+use CanisterInstallMode::Upgrade;
+use DeployArchiveResult::{Failed, Success};
+
+pub async fn deploy_archive(wasm: ByteBuf) -> DeployArchiveResult {
+    let expected_hash = if let Some(hash) = state::expected_archive_hash() {
+        hash
+    } else {
+        return Failed("archive deployment disabled".to_string());
+    };
+
+    let wasm = wasm.into_vec();
+
+    let (archive_canister, install_mode, wasm_hash_verified) = match state::archive_state() {
+        NotCreated => match create_archive(&wasm).await {
+            Ok(archive) => (archive, Install, true),
+            Err(result) => return result,
+        },
+        CreationInProgress(timestamp) => {
+            if time() - timestamp > Duration::from_secs(24 * 60 * 60).as_nanos() as u64 {
+                // The archive has been in creation for more than a day and the creation process
+                // has likely failed thus another attempt should be made.
+                match create_archive(&wasm).await {
+                    Ok(archive) => (archive, Install, true),
+                    Err(result) => return result,
+                }
+            } else {
+                return DeployArchiveResult::CreationInProgress;
+            }
+        }
+        Created(archive_data) => {
+            let status = match archive_status(archive_data.archive_canister).await {
+                Ok(status) => status,
+                Err(message) => return Failed(message),
+            };
+            match status.module_hash {
+                None => (archive_data.archive_canister, Install, false),
+                Some(hash) if hash == expected_hash.to_vec() => {
+                    // we already have an archive with the expected module and don't need to do anything
+                    return Success(archive_data.archive_canister);
+                }
+                Some(_) => (archive_data.archive_canister, Upgrade, false),
+            }
+        }
+    };
+
+    // creating the archive verifies the wasm hash as well
+    if !wasm_hash_verified {
```

Comment:
This blind boolean is asking for trouble; why do we verify the hash in two difference places? Also, can we use types to signal whether the hash was verified? something like `newtype` would do nicely